### ISSUE TITLE
During inference, is it generated as a whole or partially?

### ISSUE THREAD

**Issue by Linghuxc** [Zain-Jiang/Speech-Editing-Toolkit#10]
Hi,
I recently discovered in experiments that the audio generated using fluentspeech is also different from the original audio in the non-modified parts. Here are the waveforms of the unmodified portion of both, with the original audio above and the generated audio below.
![image](https://github.com/Zain-Jiang/Speech-Editing-Toolkit/assets/128498268/1c4286c3-8aed-44e1-ab93-a26c9c14bf54)

However, in the fluentspeech paper, it is stated that the masked part is generated in Reverse Diffusion, and the non-mask part has not been changed.Here is the figure in the paper.
![image](https://github.com/Zain-Jiang/Speech-Editing-Toolkit/assets/128498268/2168ef9d-8153-4346-8fdd-74b18f6f4e73)

Is the generation generated as a whole or partially?

**Comment by kelvinqin** [comment]
Hi, Linghuxc,
I have the same finding --- the output of spec_denoiser.py was generated frame by frame I think, but on the demo page, the output apparently was concated by modified and unmodified part (https://speechai-demo.github.io/FluentSpeech/),

Kelvin

**Comment by danablend** [comment]
Hello, I agree. It actually sounds like the entire clip (including reference clip aka ground truth) is re-generated to fix the issue when transitioning from original sound to AI sound back to original sound.

I understand the author of this library solved this in Mega-TTS with a VQGAN-based acoustic model trained on a lot of different speech to generalize well, similar to HiFi-GAN, but for spectrogram generation. Perhaps we may be lucky to see such solutions in the future.

To my understanding, the method used in FluentSpeech involves running the original GT through the vocoder to generate a new "reference GT" which acts as GT in the demos. Then the modified audio is made by applying the masking to the original GT and using reverse diffusion to edit the speech, then run it through the vocoder at the end. Since both GT and the modified audio goes through the vocoder at the end, they sound very very similar! I may be mistaken, but it looks like this is the method used in the code:

inference/tts/spec_denoiser.py
```
def forward_model(self, inp):
        # ....code

        with torch.no_grad():
            output = self.model(edited_txt_tokens, time_mel_masks=time_mel_masks, mel2ph=edited_mel2ph, spk_embed=sample['spk_embed'],
                       ref_mels=ref_mels, f0=edited_f0, uv=edited_uv, energy=None, infer=True, use_pred_pitch=True)
            mel_out = output['mel_out'] * time_mel_masks + ref_mels * (1-time_mel_masks)
            wav_out = self.run_vocoder(mel_out) # <--- run entire output wav through vocoder
            wav_gt = self.run_vocoder(sample['mel']) # <--- run GT through vocoder
            # item_name = sample['item_name'][0]
            # np.save(f'inference_acl/mel2ph/{item_name}',output['mel2ph'].cpu().numpy()[0])

        wav_out = wav_out.cpu().numpy()
        wav_gt = wav_gt.cpu().numpy()
        mel_out = mel_out.cpu().numpy()
        mel_gt = sample['mel'].cpu().numpy()
        masked_mel_out = ref_mels.cpu().numpy()
        masked_mel_gt = (sample['mel'] * time_mel_masks_orig[:, :, None]).cpu().numpy()

        return wav_out[0], wav_gt[0], mel_out[0], mel_gt[0], masked_mel_out[0], masked_mel_gt[0]
```

Perhaps fine tuning the vocoder on the speaker will make it sound more natural? To my knowledge, HiFi-GAN generalizes well even without fine tuning, but performs even better when fine tuned to the speaker.

**Comment by Linghuxc** [comment]
> Hello, I agree. It actually sounds like the entire clip (including reference clip aka ground truth) is re-generated to fix the issue when transitioning from original sound to AI sound back to original sound.
> 
> I understand the author of this library solved this in Mega-TTS with a VQGAN-based acoustic model trained on a lot of different speech to generalize well, similar to HiFi-GAN, but for spectrogram generation. Perhaps we may be lucky to see such solutions in the future.
> 
> To my understanding, the method used in FluentSpeech involves running the original GT through the vocoder to generate a new "reference GT" which acts as GT in the demos. Then the modified audio is made by applying the masking to the original GT and using reverse diffusion to edit the speech, then run it through the vocoder at the end. Since both GT and the modified audio goes through the vocoder at the end, they sound very very similar! I may be mistaken, but it looks like this is the method used in the code:
> 
> inference/tts/spec_denoiser.py
> 
> ```
> def forward_model(self, inp):
>         # ....code
> 
>         with torch.no_grad():
>             output = self.model(edited_txt_tokens, time_mel_masks=time_mel_masks, mel2ph=edited_mel2ph, spk_embed=sample['spk_embed'],
>                        ref_mels=ref_mels, f0=edited_f0, uv=edited_uv, energy=None, infer=True, use_pred_pitch=True)
>             mel_out = output['mel_out'] * time_mel_masks + ref_mels * (1-time_mel_masks)
>             wav_out = self.run_vocoder(mel_out) # <--- run entire output wav through vocoder
>             wav_gt = self.run_vocoder(sample['mel']) # <--- run GT through vocoder
>             # item_name = sample['item_name'][0]
>             # np.save(f'inference_acl/mel2ph/{item_name}',output['mel2ph'].cpu().numpy()[0])
> 
>         wav_out = wav_out.cpu().numpy()
>         wav_gt = wav_gt.cpu().numpy()
>         mel_out = mel_out.cpu().numpy()
>         mel_gt = sample['mel'].cpu().numpy()
>         masked_mel_out = ref_mels.cpu().numpy()
>         masked_mel_gt = (sample['mel'] * time_mel_masks_orig[:, :, None]).cpu().numpy()
> 
>         return wav_out[0], wav_gt[0], mel_out[0], mel_gt[0], masked_mel_out[0], masked_mel_gt[0]
> ```
> 
> Perhaps fine tuning the vocoder on the speaker will make it sound more natural? To my knowledge, HiFi-GAN generalizes well even without fine tuning, but performs even better when fine tuned to the speaker.

Yes! I think you are right!

**Comment by Zain-Jiang** [comment]
@danablend Thank you for your accurate and detailed response.
